### PR TITLE
Latency test: Replace init() with getters

### PR DESCRIPTION
**Pull request by shajmakh**
In a goal to not cache the values of the environment variables, we replaced the removed init() and replaced its functionality with getters that are being called inside the main Describe

### CODE REVIEW

**Comment by shajmakh** [comment]
@fromanirh @MarSik

File: functests/4_latency/latency.go [hunk 1]
```diff
@@ -255,6 +228,65 @@ var _ = Describe("[performance] Latency Test", func() {
 	})
 })
 
+//read LATENCY_TEST_RUN environment var and parse it if set
+func getLatencyTestRun() {
+	latencyTestRunEnv := os.Getenv("LATENCY_TEST_RUN")
+	if latencyTestRunEnv != "" {
+		if latencyTestRunEnv == "true" {
+			latencyTestRun = true
```
Comment (MarSik) [inline functests/4_latency/latency.go:236]: Thsi is still caching. The getter should just parse the value and return true/false. The `latencyTestRun` variable should be removed

File: functests/4_latency/latency.go [hunk 2]
```diff
@@ -255,6 +228,65 @@ var _ = Describe("[performance] Latency Test", func() {
...
+		}
+	}
+}
+
+//read LATENCY_TEST_RUNTIME environment var and parse it if set
+func getLatencyTestRuntime() {
+	latencyTestRuntimeEnv := os.Getenv("LATENCY_TEST_RUNTIME")
+	if latencyTestRuntimeEnv != "" {
+		latencyTestRuntime = latencyTestRuntimeEnv
```
Comment (MarSik) [inline functests/4_latency/latency.go:245]: This is still caching. The getter should just parse the value and return the string. The `latencyTestRuntime` variable should be removed

File: functests/4_latency/latency.go [hunk 3]
```diff
@@ -255,6 +228,65 @@ var _ = Describe("[performance] Latency Test", func() {
...
+}
+
+//read LATENCY_TEST_DELAY environment var and parse it if set
+func getLatencyTestDelay() {
+	latencyTestDelayEnv := os.Getenv("LATENCY_TEST_DELAY")
+	if latencyTestDelayEnv != "" {
+		var err error
+		if latencyTestDelay, err = strconv.Atoi(latencyTestDelayEnv); err != nil {
+			klog.Fatalf("the environment variable LATENCY_TEST_DELAY has incorrect value %q", latencyTestDelayEnv)
+		}
```
Comment (MarSik) [inline functests/4_latency/latency.go:256]: Same here, instead of saving the parsed value, return it.

File: functests/4_latency/latency.go [hunk 4]
```diff
@@ -255,6 +228,65 @@ var _ = Describe("[performance] Latency Test", func() {
...
+
+//read LATENCY_TEST_CPUS environment var and parse it if set
+func getLatencyTestCpus() {
+	latencyTestCpusEnv := os.Getenv("LATENCY_TEST_CPUS")
+	if latencyTestCpusEnv != "" {
+		var err error
+		if latencyTestCpus, err = strconv.Atoi(latencyTestCpusEnv); err != nil {
+			klog.Fatalf("the environment variable LATENCY_TEST_CPUS has incorrect value %q", latencyTestCpusEnv)
+		}
+	}
```
Comment (MarSik) [inline functests/4_latency/latency.go:268]: Same here, instead of saving the parsed value, return it.

File: functests/4_latency/latency.go [hunk 5]
```diff
@@ -255,6 +228,65 @@ var _ = Describe("[performance] Latency Test", func() {
...
+
+//read MAXIMUM_LATENCY environment var and parse it if set
+func getMaxLatency() {
+	unifiedMaxLatencyEnv := os.Getenv("MAXIMUM_LATENCY")
+	if unifiedMaxLatencyEnv != "" {
+		var err error
+		if maximumLatency, err = strconv.Atoi(unifiedMaxLatencyEnv); err != nil {
+			klog.Fatalf("the environment variable MAXIMUM_LATENCY has incorrect value %q", unifiedMaxLatencyEnv)
+		}
+	}
```
Comment (MarSik) [inline functests/4_latency/latency.go:279]: Same here, instead of saving the parsed value, return it.

File: functests/4_latency/latency.go [hunk 6]
```diff
@@ -255,6 +228,65 @@ var _ = Describe("[performance] Latency Test", func() {
...
+}
+
+func readAndParseEnvVar() {
```
Comment (MarSik) [inline functests/4_latency/latency.go:282]: This is not needed at all, each test should just call the getter directly

File: functests/4_latency/latency.go [hunk 7]
```diff
@@ -375,13 +407,6 @@ func extractLatencyValues(testName string, exp string, node *corev1.Node) string
 // MAXIMUM_LATENCY: unified expected maximum latency for all tests
 func setMaximumLatencyValue(testName string) error {
```
Comment (MarSik) [inline functests/4_latency/latency.go:408]: I think this should be converted into a getter as well to get rid of the `maximumLatency` variable

**Review by MarSik — CHANGES_REQUESTED** [review]

File: functests/4_latency/latency.go (hunk 2 above)
Comment (shajmakh) [reply functests/4_latency/latency.go:245]: @MarSik should we still check the condition if the value is not empty?

Comment (shajmakh) [reply functests/4_latency/latency.go:245]: because the intention in the initial code was to set new value into the vars if the new value is not empty , otherwise to keep it default

Comment (fromanirh) [reply functests/4_latency/latency.go:245]: ti seems a good opportunity to use `LookupEnv` + `ParseBool`:
```
func getLatencyTestRun() {
        // TODO: log if/where needed
        if envValue, ok := os.LookupEnv("LATENCY_TEST_RUN"); ok {
            if value, err := strconv.ParseBool(envValue); err == nil {
                 return value // this could be true or false, but it explicitely signal the intent of the user
            }    
        }
        // something was wrong or missing, stick with the builtin default
      return false
}
```

File: functests/4_latency/latency.go (hunk 3 above)
Comment (shajmakh) [reply functests/4_latency/latency.go:263]: fixed, thanks.

File: functests/4_latency/latency.go (hunk 4 above)
Comment (shajmakh) [reply functests/4_latency/latency.go:278]: fixed, thanks.

File: functests/4_latency/latency.go (hunk 5 above)
Comment (shajmakh) [reply functests/4_latency/latency.go:296]: fixed, thanks.

File: functests/4_latency/latency.go (hunk 1 above)
Comment (shajmakh) [reply functests/4_latency/latency.go]: fixed, thanks.

File: functests/4_latency/latency.go (hunk 6 above)
Comment (shajmakh) [reply functests/4_latency/latency.go]: fixed , thanks .

File: functests/4_latency/latency.go (hunk 7 above)
Comment (shajmakh) [reply functests/4_latency/latency.go]: fixed , thanks

File: functests/4_latency/latency.go [hunk 8]
```diff
@@ -229,62 +235,74 @@ var _ = Describe("[performance] Latency Test", func() {
...
+func getMaximumLatency(testName string) (parsedValue int, errMsg error) {
+	var err error
+	var val int
 	unifiedMaxLatencyEnv := os.Getenv("MAXIMUM_LATENCY")
 	if unifiedMaxLatencyEnv != "" {
 		var err error
-		if maximumLatency, err = strconv.Atoi(unifiedMaxLatencyEnv); err != nil {
-			klog.Fatalf("the environment variable MAXIMUM_LATENCY has incorrect value %q", unifiedMaxLatencyEnv)
+		if val, err = strconv.Atoi(unifiedMaxLatencyEnv); err != nil {
+			err = fmt.Errorf("the environment variable MAXIMUM_LATENCY has incorrect value %q", unifiedMaxLatencyEnv)
```
Comment (Tal-or) [inline functests/4_latency/latency.go:294]: You should return the error here, otherwise, you might override it later, when you're trying to parse the other env variables.

File: functests/4_latency/latency.go [hunk 9]
```diff
@@ -56,8 +55,15 @@ var _ = Describe("[performance] Latency Test", func() {
 	var workerRTNode *corev1.Node
 	var profile *performancev2.PerformanceProfile
 	var latencyTestPod *corev1.Pod
+	var err error
+	latencyTestDelay, err = getLatencyTestDelay()
```
Comment (Tal-or) [inline functests/4_latency/latency.go:59]: If I understood correctly, this PR intended to check the environment variables each time for each tool.
As @MarSik mentioned, storing the values here, won't achieve that, because the ```Describe()``` is common to all three test cases.
You can call theses ```get*()``` functions inside each ```Context()``` or set the global variables (```latencyTestDelay```, ```latencyTestRuntime```, etc.)  in the ```BeforeEach()```

File: functests/4_latency/latency.go (hunk 8 above)
Comment (fromanirh) [reply functests/4_latency/latency.go:294]: correct. In addition you should wrap errors here: https://blog.golang.org/go1.13-errors
TL;DR: use something like
```
err = fmt.Errorf("the environment variable MAXIMUM_LATENCY has incorrect value %q: %w", unifiedMaxLatencyEnv, err)
```

Comment (shajmakh) [reply functests/4_latency/latency.go:294]: fixed, thanks.

File: functests/4_latency/latency.go (hunk 9 above)
Comment (shajmakh) [reply functests/4_latency/latency.go:59]: fixed, thanks.

File: functests/4_latency/latency.go (hunk 2 above)
Comment (shajmakh) [reply functests/4_latency/latency.go]: yes good one, I used it. thanks .

File: functests/4_latency/latency.go (hunk 8 above)
Comment (Tal-or) [reply functests/4_latency/latency.go]: @fromanirh please keep me honest here, but yet we should have a ```return``` statement here. Storing the error in ```err``` won't keep it from being override later on in the function.

Comment (fromanirh) [reply functests/4_latency/latency.go]: it won't indeed

Comment (shajmakh) [reply functests/4_latency/latency.go]: After discussing the matter with Talor, I added a return statement on the first error check.Thanks.

File: functests/4_latency/latency.go [hunk 10]
```diff
@@ -235,45 +253,42 @@ var _ = Describe("[performance] Latency Test", func() {
...
-	latencyTestRunEnv := os.Getenv("LATENCY_TEST_RUN")
-	if latencyTestRunEnv != "" {
-		if latencyTestRunEnv == "true" {
-			val = true
-		}
+	if latencyTestRunEnv, ok := os.LookupEnv("LATENCY_TEST_RUN"); ok {
+		val, err = strconv.ParseBool(latencyTestRunEnv)
 	}
-	return val
+	return val, err
```
Comment (fromanirh) [inline functests/4_latency/latency.go:250]: this is good. Even better:
```
func getLatencyTestRun() (bool, error) {
	if latencyTestRunEnv, ok := os.LookupEnv("LATENCY_TEST_RUN"); ok {
		return strconv.ParseBool(latencyTestRunEnv)
	}
	return defaultTestRun, nil
}
```

File: functests/4_latency/latency.go [hunk 11]
```diff
@@ -235,45 +253,42 @@ var _ = Describe("[performance] Latency Test", func() {
...
 func getLatencyTestDelay() (parsedValue int, errMsg error) {
 	var err error
-	var val int
-	latencyTestDelayEnv := os.Getenv("LATENCY_TEST_DELAY")
-	if latencyTestDelayEnv != "" {
+	val := defaultTestDelay
+	if latencyTestDelayEnv, ok := os.LookupEnv("LATENCY_TEST_DELAY"); ok {
 		if val, err = strconv.Atoi(latencyTestDelayEnv); err != nil {
-			err = fmt.Errorf("the environment variable LATENCY_TEST_DELAY has incorrect value %q", latencyTestDelayEnv)
+			err = fmt.Errorf("the environment variable LATENCY_TEST_DELAY has incorrect value %q: %w", latencyTestDelayEnv, err)
```
Comment (fromanirh) [inline functests/4_latency/latency.go:267]: this is good. Even better: make it look like `getLatencyTestRuntime`. The skeleton code is the same, it should be doable.

**Review by fromanirh — COMMENTED** [review]
looks good but please squash all the commits into the first one, so we can actually check and do a final pass.

File: functests/4_latency/latency.go (hunk 11 above)
Comment (shajmakh) [reply functests/4_latency/latency.go:267]: @fromanirh  Should we preserve the modified error message below ?
fmt.Errorf("the environment variable LATENCY_TEST_DELAY has incorrect value %q: %w", latencyTestDelayEnv, err)

**Review by fromanirh — COMMENTED** [review]
codewise looks nice, please check the rebase is correct.
I'll have a final pass later today.

**Comment by shajmakh** [comment]
> codewise looks nice, please check the rebase is correct.
> I'll have a final pass later today.
Thank you, the rebase is correct now.

**Comment by fromanirh** [comment]
the CI failure is not a flake:
```
 Verifying that all code is committed after updating deps and formatting and generating code
hack/verify-generated.sh
uncommitted generated files. run 'make generate' and commit results.
 M functests/4_latency/latency.go
make: *** [Makefile:297: verify] Error 1
{"component":"entrypoint","error":"wrapped process failed: exit status 2","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-08-05T07:53:47Z"} 
```

File: functests/4_latency/latency.go [hunk 12]
```diff
@@ -255,6 +240,71 @@ var _ = Describe("[performance] Latency Test", func() {
 	})
 })
 
+//read LATENCY_TEST_RUN environment var and parse it if set
```
Comment (fromanirh) [inline functests/4_latency/latency.go:243]: minor: this is not a good godoc comment - but it's not relevant because the function is not exported. A good godoc comment should be in the form
```
// Foobar does...
func Foobar() {}
```
note the comment starts with the same name as the function

**Review by fromanirh — COMMENTED** [review]
/lgtm

File: functests/4_latency/latency.go (hunk 12 above)
Comment (shajmakh) [reply functests/4_latency/latency.go:243]: thanks for the info will consider it next time. I will remove it.

Comment (fromanirh) [reply functests/4_latency/latency.go:243]: no need, let's consider next time.

Comment (shajmakh) [reply functests/4_latency/latency.go]: got that !
